third_party/forked/prometheus/prometheus/promql/prometheus_server_utils.go

```diff
@@ -0,0 +1,162 @@
+	ensureMonotonic(buckets)
+
+	if len(buckets) < 2 {
+		return math.NaN()
+	}
+
+	rank := q * buckets[len(buckets)-1].Count
+	b := sort.Search(len(buckets)-1, func(i int) bool { return buckets[i].Count >= rank })
+
+	if b == len(buckets)-1 {
+		return buckets[len(buckets)-2].UpperBound
+	}
+	if b == 0 && buckets[0].UpperBound <= 0 {
+		return buckets[0].UpperBound
+	}
+	var (
+		bucketStart float64
+		bucketEnd   = buckets[b].UpperBound
+		count       = buckets[b].Count
+	)
+	if b > 0 {
+		bucketStart = buckets[b-1].UpperBound
+		count -= buckets[b-1].Count
+		rank -= buckets[b-1].Count
+	}
+	return bucketStart + (bucketEnd-bucketStart)*(rank/count)
+}
+
+// coalesceBuckets merges buckets with the same upper bound.
+//
+// The input buckets must be sorted.
+func coalesceBuckets(buckets Buckets) Buckets {
+	last := buckets[0]
+	i := 0
+	for _, b := range buckets[1:] {
+		if b.UpperBound == last.UpperBound {
+			last.Count += b.Count
+		} else {
+			buckets[i] = last
+			last = b
+			i++
+		}
+	}
+	buckets[i] = last
+	return buckets[:i+1]
+}
+
+// The assumption that bucket counts increase monotonically with increasing
+// upperBound may be violated during:
+//
+//   * Recording rule evaluation of histogram_quantile, especially when rate()
+//      has been applied to the underlying bucket timeseries.
+//   * Evaluation of histogram_quantile computed over federated bucket
+//      timeseries, especially when rate() has been applied.
+//
+// This is because scraped data is not made available to rule evaluation or
+// federation atomically, so some buckets are computed with data from the
+// most recent scrapes, but the other buckets are missing data from the most
+// recent scrape.
+//
+// Monotonicity is usually guaranteed because if a bucket with upper bound
+// u1 has count c1, then any bucket with a higher upper bound u > u1 must
+// have counted all c1 observations and perhaps more, so that c  >= c1.
+//
+// Randomly interspersed partial sampling breaks that guarantee, and rate()
+// exacerbates it. Specifically, suppose bucket le=1000 has a count of 10 from
+// 4 samples but the bucket with le=2000 has a count of 7 from 3 samples. The
+// monotonicity is broken. It is exacerbated by rate() because under normal
+// operation, cumulative counting of buckets will cause the bucket counts to
+// diverge such that small differences from missing samples are not a problem.
+// rate() removes this divergence.)
+//
+// bucketQuantile depends on that monotonicity to do a binary search for the
+// bucket with the φ-quantile count, so breaking the monotonicity
+// guarantee causes bucketQuantile() to return undefined (nonsense) results.
+//
+// As a somewhat hacky solution until ingestion is atomic per scrape, we
+// calculate the "envelope" of the histogram buckets, essentially removing
+// any decreases in the count between successive buckets.
+
+func ensureMonotonic(buckets Buckets) {
+	max := buckets[0].Count
+	for i := range buckets[1:] {
```

Comment:
This code neglects to check the final bucket? if len(buckets) is 10, then len(buckets[1:]) is 9, and i will range from 0 to 8. I'd expect the author wanted it to go from 1 to 9.

Reply:
(the final bucket is ignored in the search so it looks like this doesn't matter?)